sound/soc/sof/trace.c

```diff
@@ -166,19 +166,23 @@ int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
 	if (ret < 0) {
 		dev_err(sdev->dev,
 			"error: can't set params for DMA for trace %d\n", ret);
-		return ret;
+		goto trace_release;
 	}
 
 	ret = snd_sof_dma_trace_trigger(sdev, SNDRV_PCM_TRIGGER_START);
 	if (ret < 0) {
 		dev_err(sdev->dev,
 			"error: snd_sof_dma_trace_trigger: start: %d\n", ret);
-		return ret;
+		goto trace_release;
 	}
 
 	sdev->dtrace_is_enabled = true;
 
 	return 0;
+
+trace_release:
+	snd_sof_dma_trace_release(sdev);
```

Comment:
Would you mind looking at intel/hda-trace.c and fixing the code in the _init() function while you are at it ? there are tests on if (hda) after the indirection is made, and the tag is stored before the DMA prepare is done. This deserves a proper cleanup.

Reply:
sure, let me fix that together. I think we can change that directly like this:
```
---	if (hda && !hda->dtrace_stream) {
+++	if (!hda->dtrace_stream) {
```
@plbossart make sense?

Reply:
That was one cleanup, the other one is to make sure the tag is released if hda_dsp_trace_prepare() fails. Unless I am mistaken this error flow isn't handled.

Reply:
> That was one cleanup, the other one is to make sure the tag is released if hda_dsp_trace_prepare() fails. Unless I am mistaken this error flow isn't handled.

good, let me do that also now.